test/infer/test_gradient.py

```diff
@@ -0,0 +1,724 @@
+#          b = pyro.sample("b", dist.Categorical(probs_b[a.long()]))
+#          pyro.sample("c", dist.Normal(b.to(data.dtype), 1), obs=data)
+#
+#
+#  def guide_1(data):
+#      probs_a = pyro.param(
+#          "probs_a",
+#          lambda: torch.tensor([0.5, 0.5]),
+#      )
+#      a = pyro.sample("a", dist.Categorical(probs_a))
+#      with pyro.plate("data", len(data)) as idx:
+#          probs_b = pyro.param(
+#              "probs_b",
+#              lambda: torch.tensor(
+#                  [[[0.5, 0.5], [0.6, 0.4]], [[0.4, 0.6], [0.35, 0.65]]]
+#              ),
+#          )
+#          pyro.sample("b", dist.Categorical(Vindex(probs_b)[a.long(), idx]))
+#
+#
+#  def model_2(data):
+#      prob_b = torch.tensor([[0.3, 0.7], [0.4, 0.6]])
+#      prob_c = torch.tensor([[0.5, 0.5], [0.6, 0.4]])
+#      prob_d = torch.tensor([[0.2, 0.8], [0.3, 0.7]])
+#      prob_e = torch.tensor([[0.5, 0.5], [0.1, 0.9]])
+#      a = pyro.sample("a", dist.Categorical(torch.tensor([0.3, 0.7])))
+#      with pyro.plate("data", len(data)):
+#          b = pyro.sample("b", dist.Categorical(prob_b[a.long()]))
+#          c = pyro.sample("c", dist.Categorical(prob_c[b.long()]))
+#          pyro.sample("d", dist.Categorical(prob_d[b.long()]))
+#          pyro.sample("e", dist.Categorical(prob_e[c.long()]), obs=data)
+#
+#
+#  def guide_2(data):
+#      prob_a = pyro.param("prob_a", lambda: torch.tensor([0.5, 0.5]))
+#      prob_b = pyro.param("prob_b", lambda: torch.tensor([[0.4, 0.6], [0.3, 0.7]]))
+#      prob_c = pyro.param(
+#          "prob_c",
+#          lambda: torch.tensor([[[0.3, 0.7], [0.8, 0.2]], [[0.2, 0.8], [0.5, 0.5]]]),
+#      )
+#      prob_d = pyro.param(
+#          "prob_d",
+#          lambda: torch.tensor([[[0.2, 0.8], [0.9, 0.1]], [[0.1, 0.9], [0.4, 0.6]]]),
+#      )
+#      a = pyro.sample("a", dist.Categorical(prob_a))
+#      with pyro.plate("data", len(data)) as idx:
+#          b = pyro.sample("b", dist.Categorical(prob_b[a.long()]))
+#          pyro.sample("c", dist.Categorical(Vindex(prob_c)[b.long(), idx]))
+#          pyro.sample("d", dist.Categorical(Vindex(prob_d)[b.long(), idx]))
+
+
+@pytest.mark.parametrize(
+    "model,guide,params,data",
+    [
+        (model_0, guide_0, params_0, jnp.array([-0.5, 2.0])),
+        #  (model_1, guide_1, torch.tensor([-0.5, 2.0])),
+        #  (model_2, guide_2, torch.tensor([0.0, 1.0])),
+    ],
+)
+def test_gradient(model, guide, params, data):
+
+    # Expected grads based on exact integration
+    elbo = infer.TraceEnum_ELBO(
+        max_plate_nesting=1,  # set this to ensure rng agrees across runs
+    )
+    expected_loss_fn = lambda params: elbo.loss(
+        random.PRNGKey(0), {}, model, config_enumerate(guide), data, params
+    )
+    expected_loss, expected_grads = jax.value_and_grad(expected_loss_fn)(params)
+
+    # Actual grads averaged over num_particles
+    elbo = infer.Trace_ELBO(
+        num_particles=50000,
+    )
```

Comment:
I just read the docs for `Trace_ELBO` and it says that it doesn't support non reparametrized samplers. So I tried to use `TraceGraph_ELBO` instead here but got gradients that are shifted by 1 compared to `TraceEnum_ELBO`:

```
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0.02
E   
E   Mismatched elements: 4 / 4 (100%)
E   Max absolute difference: 1.0056443
E   Max relative difference: 0.47962338
E    x: array([[1.332346, 1.889065],
E          [3.42412 , 1.083667]], dtype=float32)
E    y: array([[2.331621, 2.889788],
E          [4.429764, 2.082466]], dtype=float32)
```

Reply:
That difference seems to be due to these lines in `TraceGraph_ELBO`:

https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/elbo.py#L687-L688